sway-core/src/semantic_analysis/type_check_analysis.rs

```diff
@@ -0,0 +1,249 @@
+//! This module handles the process of iterating through the typed AST and doing an analysis.
+//! At the moment we compute an dependency graph between typed nodes.
+
+use std::collections::{HashMap, HashSet};
+use std::fs;
+
+use petgraph::stable_graph::NodeIndex;
+use petgraph::Graph;
+use sway_error::handler::{ErrorEmitted, Handler};
+
+use crate::decl_engine::{DeclId, DeclIdIndexType, DeclRef};
+use crate::engine_threading::DebugWithEngines;
+use crate::language::ty::{self, TyImplItem, TyTraitItem};
+use crate::Engines;
+
+pub type TyNodeDepGraphNodeId = petgraph::graph::NodeIndex;
+
+#[derive(Clone, Debug)]
+pub struct TyNodeDepGraphEdge(pub String);
```

Comment:
These should be typed as an enum instead of manipulating raw strings.

Reply:
Maybe I'm misunderstanding what this is for, but I see literal instantiations of it using those strings and it looks like this is stored in the graph. Even if this is debug information it seems like pure overheat to store it in there as owned strings that are all the same instead of some enum when there's a limited set of possibilities for those messages.